oauth-common/src/main/java/io/strimzi/kafka/oauth/common/HttpUtil.java

```diff
@@ -45,6 +46,44 @@ public class HttpUtil {
     static final int DEFAULT_CONNECT_TIMEOUT = getConnectTimeout(new Config());
     static final int DEFAULT_READ_TIMEOUT = getReadTimeout(new Config());
 
+    public static <T> T doWithRetries(int retries, long retryPauseMillis, IOTask<T> task) throws ExecutionException {
+
+        if (retries < 0) {
+            throw new IllegalArgumentException("retries can't be negative");
+        }
+
+        retryPauseMillis = retryPauseMillis < 0 ? 0 : retryPauseMillis;
+
+        Exception exception;
+        int i = 0;
+        do {
+            i++;
+
+            if (i > 1 && retryPauseMillis > 0) {
+                log.debug("Pausing before retrying failed action (for {}ms)", retryPauseMillis);
+                try {
+                    Thread.sleep(retryPauseMillis);
```

Comment:
I'm not sure about the fixed pause time. If something goes wrong with an endpoint, more clients are likely to end up in the retry loop the longer the endpoint is down. A recovering Oauth server is likely to be struck with significant load. 

Because of the blocking/thread-using nature of what we're doing here, I suspect users will be inclined to stick to small pause times and smallish numbers of retries. Suppose they configure 5 retries with a 200ms pause. If the server normally sees 50 requests/s, then if it's down for > 1s when it comes back it will see those 50 requests in the first 200ms, i.e. 5 times higher than normal. If it were a JVM-based server this would happen before JIT had kicked in making the problem worse because the response time would be slower, so it would end up seeing even higher load.

So I wonder if we should apply a backoff so that the peak server load isn't as bad.

Reply:
Yeah I'm trying to avoid is for any request to idle, since the calling logic is inside the Kafka worker thread and especially on the broker side, idling might quickly exhaust the worker pool. I don't know how worker threads are distributed among listeners but if the worker pool is shared then even non-oauth listeners might end up without available worker threads. I foresee the retry count of 2 or 3 max with a low retryPause, maybe 1s max. But the dam effect for the case when authorization server is not available for a great number of requests can become a problem yes.

OTOH one should always size the authz server 10x to survive such spikes.

I'd say, my priority here is - don't break the broker. Let authorization server protect itself, maybe by refusing / dropping connections if it's under too much stress. 

Another factor to be aware of is also connect timeout and read timeout which default to 60 seconds. Without explicitly configuring different values that's another point of possible blocking for a long time.

If this is used in the standalone client app then an exponential backoff would be less of a problem from thread exhaustion POV.

Maybe we don't add exponential backoff for now. If we decide to need it in the future we can add a boolean flag that enables it, and multiply the configured retryPauseMillis x2 for each next retry.